### PR TITLE
Fix subpixel misalignment when an image dimension is not a multiple of 4

### LINKED ISSUE #15: Keypoints are misaligned by up to half a pixel when an image dimension is not a multiple of 4
Same root cause as Parskatt/dad#18: LoMa carries its own copy of the DaD detector in `src/loma/detector/dad.py`, so it shows the same behaviour and can be fixed here independently of upstream.

`to_pixel_coords` maps a score-map index `i` to `i + 0.5` on a `[0, W]` axis, so a horizontal flip maps `x` to `W - x` exactly, and detections on an image and on its mirror must agree to the subpixel. They do not, unless the dimension is a multiple of 4: the error reaches 0.5 px, following `W % 4` on x and `H % 4` on y independently.

`DaD.load_image` floors both dimensions with `(scale * W) // 8 * 8`, so that path never hits this. Any caller passing its own tensor to `detect` does, including the proposed hloc extractor in cvg/Hierarchical-Localization#495, which forwards hloc's resized image at the source aspect ratio.

### Cause

`forward_impl` upsamples the logits back through the sizes recorded by the encoder. Those are floor-halvings, so a dimension that does not halve exactly makes an `F.interpolate` ratio non-integral, and with `align_corners=False` the score map lands off the pixel grid. Only the two finest levels matter in practice, which is why the condition is `% 4` and not `% 8`:

| width | %8 | %4 | upsample ratios (fine to coarse) | biased |
|---|---|---|---|---|
| 497 | 1 | 1 | **2.0040**, 2.0000, 2.0000 | yes |
| 498 | 2 | 2 | 2.0000, **2.0081**, 2.0000 | yes |
| 499 | 3 | 3 | **2.0040**, **2.0081**, 2.0000 | yes |
| 500 | 4 | 0 | 2.0000, 2.0000, **2.0161** | no |
| 504 | 0 | 0 | 2.0000, 2.0000, 2.0000 | no |

500 and 508 are the discriminating cases: not multiples of 8, but unbiased, because only the coarsest ratio is non-integral and that level is smoothed away.

### Reproduction

```python
"""Keypoints are not flip-equivariant unless the image dimensions are multiples of 4."""

import numpy as np
import torch
from scipy.spatial import cKDTree
from scipy.stats import trim_mean
from skimage import data

from loma.detector.dad import DaD

detector = DaD(DaD.Cfg(compile=False)).eval().cuda()


def pixel_keypoints(image):
    tensor = torch.from_numpy(image.transpose(2, 0, 1)[None] / 255.0).float().cuda()
    with torch.inference_mode():
        kpts = detector.detect({"image": tensor}, num_keypoints=2048)["keypoints"][0]
    coords = detector.to_pixel_coords(kpts, image.shape[0], image.shape[1])
    return coords.cpu().numpy().astype(np.float64)


base = np.asarray(data.astronaut())[..., :3]
print(" W    W%4   x offset (expected +0.000)")
for width in range(497, 513):
    crop = np.ascontiguousarray(base[:512, :width])
    original = pixel_keypoints(crop)
    mirrored = pixel_keypoints(np.ascontiguousarray(crop[:, ::-1]))
    mirrored[:, 0] = width - mirrored[:, 0]
    distance, index = cKDTree(original).query(mirrored, k=1)
    near = distance < 2.0
    offset = (mirrored[near] - original[index[near]])[:, 0]
    flag = "" if abs(trim_mean(offset, 0.2)) < 0.05 else "   <-- biased"
    print(f" {width}   {width % 4}     {trim_mean(offset, 0.2):+.3f}{flag}")
```

Output on a Tesla T4, torch 2.14.0+cu130, LoMa at 8fb59c4:

```
 W    W%4   x offset (expected +0.000)
 497   1     -0.366   <-- biased
 498   2     -0.087   <-- biased
 499   3     -0.481   <-- biased
 500   0     -0.002
 501   1     -0.388   <-- biased
 502   2     -0.103   <-- biased
 503   3     -0.495   <-- biased
 504   0     +0.004
 505   1     -0.369   <-- biased
 506   2     -0.088   <-- biased
 507   3     -0.487   <-- biased
 508   0     -0.001
 509   1     -0.386   <-- biased
 510   2     -0.103   <-- biased
 511   3     -0.488   <-- biased
 512   0     +0.003
```

Sweeping the height instead gives the same pattern on y (`-0.432`, `-0.117`, `-0.549` for `H % 4` of 1, 2, 3). The axes are independent: across 16 widths the y residual holds at `-0.008`, and across 16 heights the x residual holds at `+0.005`.

### Suggested fix

Padding the bottom-right up to a multiple of 4, then cropping the score map back, removes the bias at every size and leaves already-correct sizes bit-identical. Padding to 8 instead would also make the coarsest halving exact, but that was measured to move keypoints by only 0.0025 px, so 4 is the smallest sufficient modulus. Overhead is under 0.2% of a `detect` call (0.04 ms of ~23 ms on a T4). #16 does that, with a test covering one size per residue class on both axes.

### PR DESCRIPTION
**Pull request by f-dy**
Closes #15

Same root cause as upstream Parskatt/dad#18, fixed there by Parskatt/dad#17 with the identical patch. LoMa carries its own copy of the detector in `src/loma/detector/dad.py`, so this can land independently of upstream.

Keypoint coordinates are misaligned by up to half a pixel when an image dimension is not a multiple of 4.

### Cause

`forward_impl` upsamples the logits back through the sizes recorded by the encoder. Those are floor-halvings, so a dimension that does not halve exactly makes an `F.interpolate` ratio non-integral, and with `align_corners=False` the score map lands off the pixel grid.

Only the two finest levels move a keypoint measurably, so the condition is `% 4`, not `% 8`:

| width | %8 | %4 | upsample ratios (fine to coarse) | biased |
|---|---|---|---|---|
| 497 | 1 | 1 | **2.0040**, 2.0000, 2.0000 | yes |
| 498 | 2 | 2 | 2.0000, **2.0081**, 2.0000 | yes |
| 499 | 3 | 3 | **2.0040**, **2.0081**, 2.0000 | yes |
| 500 | 4 | 0 | 2.0000, 2.0000, **2.0161** | no |
| 508 | 4 | 0 | 2.0000, 2.0000, **2.0161** | no |
| 504 | 0 | 0 | 2.0000, 2.0000, 2.0000 | no |

500 and 508 are the discriminating cases: not multiples of 8, but unbiased, because only the coarsest ratio is non-integral and that level is smoothed away.

`load_image` floors both dimensions with `(scale * W) // 8 * 8`. A multiple of 8 is also a multiple of 4, so that path satisfies the condition and never exposes this. A caller passing its own tensor to `detect` gets no such guarantee.

### Fix

Pad the bottom-right up to a multiple of 4, then crop the score map back. Padding after the origin leaves the top-left alignment unchanged, so sizes that were already correct produce identical output.

Padding to 8 instead would also make the coarsest halving exact. Measured against padding to 4, that moves keypoints by **0.0025 px** and is bit-identical wherever both round to the same size, so 4 is used as the smallest sufficient modulus. Overhead is under 0.2% of a `detect` call (0.04 ms of ~23 ms on a T4).

### Measured

Trimmed-mean flip-equivariance residual, which must be 0. `to_pixel_coords` maps index `i` to `i + 0.5` on a `[0, W]` axis, so a flip maps `x` to `W - x` exactly. One image cropped to each size, so content is constant.

| dim % 4 | x before | x after | y before | y after |
|---|---|---|---|---|
| 0 | -0.002 | +0.003 | -0.007 | -0.007 |
| 1 | **-0.366** | +0.007 | **-0.432** | -0.007 |
| 2 | **-0.087** | +0.005 | **-0.117** | -0.010 |
| 3 | **-0.481** | +0.008 | **-0.549** | -0.010 |

x follows `W % 4` and y follows `H % 4` independently: across 16 widths the y residual stays at -0.008, and across 16 heights the x residual stays at +0.005.

More keypoints also survive the flip, since they no longer move off their true position (width 503: 1244 to 1541 correspondences).

### Tests

`tests/test_flip_equivariance.py` covers one size per residue class on both axes. Without the fix 12 of 16 cases fail and the 4 that pass are exactly the `% 4 == 0` sizes, so it is not vacuous. With the fix all 16 pass, and `tests/test_smoke.py` is unchanged at 3 passed.

Draft because you may prefer the padding in `detect` rather than `forward_impl`.

### CODE REVIEW

**Comment by f-dy** [comment]
Superseded by #16, which has the same change with a clean history.
